lib/component/catalog_table.es6.js

```diff
@@ -0,0 +1,137 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+function catalogTableController($scope, api) {
+  $scope.apiCatalogMatrix = {};
+  $scope.showRows = {};
+  $scope.currentPage = 0;
+  $scope.gap = 5;
+  $scope.itemPerPage = 15;
+  let ctrl = this;
+  let apiMatrix = api.matrix;
+  let setPageLength = function() {
+    $scope.currentPage = 0;
+    $scope.gap = 5;
+    $scope.itemPerPage = 15;
+    $scope.pageLength = Math.ceil(Object.keys($scope.apiCatalogMatrix).length
+      / $scope.itemPerPage);
+  };
+
+  let displayMatrix = function(matrix) {
+    $scope.apiCatalogMatrix = matrix;
+    setPageLength();
+    $scope.$apply();
+  };
+
+  // An helper function to get browserKeys from an
+  // array of browser objects.
+  let getBrowserKeys = function(browsers) {
+    return browsers.map((browser) => {
+      return browser.browserKey;
+    });
+  };
+
+  // Update matrix and view when browsers was changed.
+  ctrl.$onChanges = function(changes) {
+    if (changes.browsers && changes.browsers.currentValue) {
+      let browserKeys = getBrowserKeys(changes.browsers.currentValue);
+      apiMatrix.toMatrix(browserKeys, {
+        browserOptions: ctrl.browserOptions,
+        numAvailable: ctrl.numAvailable,
+      }).then(displayMatrix);
+    }
+  };
+  // Get interface names from itemPerPage * current page
+  // to itemPerPage * (currentPage + 1)
+  $scope.getInterfaceRange = function() {
+    return Object.keys(this.apiCatalogMatrix).sort().slice(
+      $scope.currentPage * $scope.itemPerPage,
+      ($scope.currentPage + 1) * $scope.itemPerPage);
+  };
+  $scope.showCatalog = function(interfaceName) {
+    if ($scope.showRows[interfaceName]) {
+      $scope.showRows[interfaceName] = false;
+    } else {
+      $scope.showRows[interfaceName] = true;
+    }
+  };
+  // When remove browser button clicked, notify
+  // parent component to delete this browser from browser list.
+  $scope.removeBrowser = function($event, browser) {
+    $event.stopPropagation();
+    ctrl.onDeleteBrowser({browser});
+  };
+  // Get page range from currentPage - 2 to currentPage + 2.
+  $scope.range = function(size, current) {
+    let ret = [];
+    let start = current - 2;
+    if (start < 0) start = 0;
+    let end = start + $scope.gap;
+    if (end > size) {
+      end = size;
+      start = size - $scope.gap;
+      if (start < 0) start = 0;
+    }
+    for (let i = start; i < end; i++) {
+      ret.push(i);
+    }
+    return ret;
+  };
+  // Set page when navigate to a new page.
+  $scope.setPage = function(p) {
+    if ($scope.currentPage === undefined) return;
```

Comment:
`if (typeof $scope.currentPage === 'undefined') return;`

That's safe against redefinition of `undefined`.